Class law changes

garyb opened the pull request:
Resolves #153, resolves #132, resolves #161

@hdgarrood perhaps you could take a look at this one?

garyb committed:
Add recommendation of equality

Somewhat resolves #153

garyb committed:
Require `DivisionRing` for `Field`

Resolves #132

LiamGoodacre commented on src/Data/Eq.purs:
```diff
@@ -18,6 +18,9 @@ import Data.Void (Void)
 -- | values that should be equivalent are not.
+-- |
+-- | It is generally recommended to declare `Eq` instances where
+-- | `x == y && f x == f y` but not a strict requirement.
```
Should this `&&` be `implies`?

Also perhaps this should be more explicit that the two `==` are potentially using different instances.

Is it worth giving a short example where this isn't true? Like:
```
newtype Identical a = Identical a
unidentical (Identical a) = a
instance eqIdentical :: Eq (Identical a) where eq _ _ = true
```
This `Eq` instance isn't isomorphic to the `Eq` of the underlying type.  Therefore, `unidentical` is a function that lets you recover the distinction between values.  If `f ~ unidentical` then ```x == y `implies` f x == f y``` is false.

garyb replied:
Ah yeah, `implies`, or just `(x == y) == (f x == f y)` I guess.

That example `Identical` example is perfectly correct, but perhaps we could come up with one that seems more "useful" :smile:. I'll have a think, as it still needs to be trivial enough that it requires not much more than the definitions here.

LiamGoodacre replied:
Well, in general `f x == f y` doesn't need to imply `x == y`.  That would be a requirement of injectivity for `f`?

LiamGoodacre replied:
Unless you're saying "for all injective `f`, the following should be true...".

garyb replied:
Yeah, right again! `f = const unit` is going to violate most instances of that. I guess "injective `f`" can be included in the description.

LiamGoodacre replied:
And then we'd possibly need to say what kind of injectivity we're talking about, because if it's with respect to `Eq`, then `Identical`/`unidentical` aren't valid `f`s -  because the `Eq` isn't iso to the inner type's `Eq`.  That is, the equality semantics need to be talking "in the same language" to make sense.
If it's injective with respect to definitional equality, then those functions are injective. But that messes things up 😅.

Perhaps a less wordy way to say all this is:
> An `Eq t` instance should ideally be consistent with the `Eq` instances of everything that can be injected into and projected out of `t`, should those instances exist.

LiamGoodacre replied:
Kmett said it better: "structural and reflexive" 😸 
https://www.reddit.com/r/haskell/comments/87mf4v/is_it_considered_acceptable_for_eq_instances_to/

hdgarrood commented:
I think it might be best to leave the first line as is, especially since we are adding the recommendation that `x == y` should imply `f x == f y`. If the class were intended just for any equivalence relation, then we wouldn’t be adding that recommendation. Also saying it’s for ‘types which

hdgarrood commented:
Oops.

....‘for types which support an equivalence relation’ doesn’t really tell you anything - since `const (const true)` is an equivalence relation which could be used for any type.

garyb commented:
@hdgarrood I just typed a bunch of stuff but then actually read what you wrote properly :wink:

Reverting the change to the first part is fine by me. I changed it because the laws we require are just those of an equivalence relation, but you're right that with the extra recommendation that is a bit strange, so :+1:.

garyb committed:
Revise `Eq` description again

garyb commented:
I've done the implementation changes for `Int` mod/div using the E-definition, but could use a hand with revising the laws accordingly (I didn't have much luck finding that paper you mentioned in #161 with the definitions in @hdgarrood).

hdgarrood commented:
The laws should stay the same, it’s `div` which needs to change - I think the current `div` is the T-definition. I think ideally the tests should check all 4 combinations of positive and negative dividends and divisors.

hdgarrood reviewed (commented):
I’ve just realised this `mod` is incorrect. We should have ``a `mod` (-b) == a `mod` b``, for the E-definition, I think. I’ll have a think about how best to implement the E-definition of `div` and `mod`.

I think we should also provide `quot :: Int -> Int -> Int` and `rem :: Int -> Int -> Int` which would be the same as the current `div` and `mod` and ideally would have inlining rules, in order to assuage people who are worried about perf, and/or for ease of porting code originally written in other languages.

hdgarrood commented:
Also, I think having the recommendation say that `x == y` should imply `f x == f y` would be closer to what we want. Consider `Map`, for instance: the `Eq (Map k v)` instance definitely isn't structural, but it is the 'correct' instance for that type, and it does follow the rule that `x == y` implies `f x == f y` for pretty much all functions `f` exported by that module.

garyb committed:
Remove intDiv/intMod/quot/rem

- `intDiv` / `intMod` can be tested normally now the compiler has been
updated
- `quot` / `rem` will go in `Data.Int`

garyb commented:
I revised the `Eq` description again to use something Liam mentioned, it avoids the structural mention and instead phrases it in terms of things that can be injected/projected from the type, which I think is still sensible interpretation of what happens for `Map`. Plus I like that it's a bit more vague :wink: given that it's a guideline rather than a law, and ``(x == y) `implies` (f x == f y)`` looks more law-y.

garyb commented:
Any thoughts on the current description @hdgarrood / @LiamGoodacre? I could just revert the change to `Eq` and leave it as it was as another alternative :smile:

LiamGoodacre commented:
Yeah let's revert the Eq description and get the rest of the changes in.  We can continue the discussion of Eq later if needed.

garyb commented:
Ok cool. Closing this one as it's a bit of a mess! (Since we did `Int` changes elsewhere)
